### DISCUSSION TITLE
Slug helper removes slashes

### DISCUSSION THREAD

**Discussion by chimit** [laravel/framework#32999] [General]
- Laravel Version: 7.4.0
- PHP Version: 7.4

### Description:
The slug helper removes slashes instead of replacing them with a dash.

Comments from `Illuminate/Support/Str.php`:
```
// Remove all characters that are not the separator, letters, numbers, or whitespace.
...
// Replace all separator characters and whitespace by a single separator
```
I guess slash is a separator symbol, therefore it shouldn't be removed but replaced by a separator.

### Steps To Reproduce:
```
php artisan tinker
>>> Str::slug('Government/Academic')
=> "governmentacademic"
```

**Comment by driesvints** [comment]
Bit opinionated I guess. Moving to discussions. I doubt this will ever make it in because of the breaking change.

**Reply by sandcore-dev** [reply]
I had the expectation that the slash would be replaced with a dash, too.

An optional parameter to replace instead of remove would be a non-breaking change, wouldn't it?

**Reply by joelmellon** [reply]
It's not just slashes, it's pretty much all non-alphanumeric characters. Underscores are replaced as expected since the regex interpreter considers them `\w`ord characters.

```php
>>> Str::slug('son_of_a');
=> "son-of-a"
>>> Str::slug('son.of.a');
=> "sonofa"
>>> Str::slug('son/of/a');
=> "sonofa"
>>> Str::slug('son+of+a');
=> "sonofa"
```

> An optional parameter to replace instead of remove would be a non-breaking change, wouldn't it?

Unfortunately, a "breaking change" is anything that changes behavior/logic, and isn't a bug fix. So any time functionality is changed/replaced (as in your use of "instead of") it's a breaking change in that it could/would "break" people's code even if they didn't make any changes.

Non-breaking changes are additions to code, that have zero impact on existing code, ie. in order to use them, you have to add code.

Regardless, I think this change would be worth it in a major release where breaking changes are expected, and a simple note can be added to the change log. Way worse breaking changes have occurred in major releases.

---

Until this is fixed, I'd look at making your own [Str macro](https://www.larashout.com/laravel-macros-extending-laravels-core-classes#toc-creating-a-laravel-macro), eg.

```php
Str::macro('properSlug', function ($str) {
    return Str::slug(preg_replace('/[^\w\d]+/', '-', $str));
});

echo Str::properStr('test/test');  // outputs 'test-test'
```

I haven't tested this, FYI. It also circumvents some of Str::slug()'s functionality, eg. this doesn't replace `@`'s with `at`.

**Reply by Oussama-Tn** [reply]
> ```
> Str::macro('properSlug', function ($str) {
>     return Str::slug(preg_replace('/[^\w\d]+/', '-', $str));
> });
> 
> echo Str::properStr('test/test');  // outputs 'test-test'
> ```
> 
> I haven't tested this, FYI. It also circumvents some of Str::slug()'s functionality, eg. this doesn't replace `@`'s with `at`.

Maybe we can add default args to the macro:
```
Str::macro('properSlug', function ($str, $separator = '-', $language = 'en') {
        $title = preg_replace('/[^\w\d]+/', $separator, $str);
        return Str::slug($title, $separator, $language);
});
```
